### PR TITLE
DM-46311: Explicitly control accepts to avoid maxing out open files.

### PR DESCRIPTION
**Pull request by ktlim**

### CODE REVIEW

File: python/s3daemon/s3daemon.py [hunk 1]
```diff
@@ -61,30 +63,29 @@
...
     Parameters
     ----------
     client : `S3`
         The S3 client to use to talk to the server.
-    reader : `asyncio.StreamReader`
-        A stream connected to the socket to read the filename/destination pair.
-    writer : `asyncio.StreamWriter`
-        A stream connected to the socket to write back status information.
+    conn : `socket.Socket`
+        The socket connected to the client.
```
Comment (timj) [inline python/s3daemon/s3daemon.py:74]: Should this document that the connection is closed in this function?

File: python/s3daemon/s3daemon.py [hunk 2]
```diff
@@ -61,30 +63,29 @@
...
         with open(filename, "rb") as f:
             await client.put_object(Body=f, Bucket=bucket, Key=key)
-            writer.write(b"Success")
-            log.info("%f %f sec - %s", start, time.time() - start, filename)
+            conn.send(b"Success")
+            log.info("%f %f sec - %s", start, time.time() - start, key)
     except Exception as e:
-        writer.write(bytes(repr(e), "UTF-8"))
-        log.exception("%f %f sec - %s", start, time.time() - start, filename)
+        conn.send(bytes(repr(e), "UTF-8"))
```
Comment (timj) [inline python/s3daemon/s3daemon.py:86]:
```suggestion
        conn.send(repr(e).encode())
```
? or `conn.send(repr(e).encode("UTF-8"))`?

File: python/s3daemon/s3daemon.py [hunk 3]
```diff
@@ -61,30 +63,29 @@
...
 
     Parameters
     ----------
     client : `S3`
         The S3 client to use to talk to the server.
-    reader : `asyncio.StreamReader`
-        A stream connected to the socket to read the filename/destination pair.
-    writer : `asyncio.StreamWriter`
-        A stream connected to the socket to write back status information.
+    conn : `socket.Socket`
```
Comment (timj) [inline python/s3daemon/s3daemon.py:73]: `pydoc socket.Socket` doesn't return anything for me? Is this the right class name? Maybe `socket.socket`?

File: python/s3daemon/s3daemon.py [hunk 4]
```diff
@@ -61,30 +63,29 @@
...
+        The socket connected to the client.
     """
-    filename, dest = (await reader.readline()).decode("UTF-8").rstrip().split(" ")
+    filename, dest = conn.recv(4096).decode("UTF-8").rstrip().split(" ")
```
Comment (timj) [inline python/s3daemon/s3daemon.py:76]: So previously `readline` was never going to return more than 4096 characters anyhow?

File: python/s3daemon/s3daemon.py [hunk 5]
```diff
@@ -61,30 +63,29 @@
 log.setLevel(logging.INFO)
 
 
-async def handle_client(client, reader, writer):
+async def handle_client(client, conn):
     """Handle a client connection to the server socket.
 
     Parameters
     ----------
     client : `S3`
```
Comment (timj) [inline python/s3daemon/s3daemon.py:71]: I can't find an `S3` class in the source code.

File: python/s3daemon/s3daemon.py [hunk 6]
```diff
@@ -61,30 +63,29 @@
...
     start = time.time()
     # ignore the alias
     _, bucket, key = dest.split("/", maxsplit=2)
     try:
         with open(filename, "rb") as f:
             await client.put_object(Body=f, Bucket=bucket, Key=key)
-            writer.write(b"Success")
-            log.info("%f %f sec - %s", start, time.time() - start, filename)
+            conn.send(b"Success")
+            log.info("%f %f sec - %s", start, time.time() - start, key)
```
Comment (timj) [inline python/s3daemon/s3daemon.py:84]: If you can use `lsst.utils.timer` then the `time_this` context manager simplifies this logging a lot.

File: python/s3daemon/s3daemon.py [hunk 7]
```diff
@@ -61,30 +63,29 @@
...
+    filename, dest = conn.recv(4096).decode("UTF-8").rstrip().split(" ")
     start = time.time()
     # ignore the alias
     _, bucket, key = dest.split("/", maxsplit=2)
     try:
         with open(filename, "rb") as f:
             await client.put_object(Body=f, Bucket=bucket, Key=key)
-            writer.write(b"Success")
-            log.info("%f %f sec - %s", start, time.time() - start, filename)
+            conn.send(b"Success")
```
Comment (timj) [inline python/s3daemon/s3daemon.py:83]: No need to include `f` in the success message?

**Review by timj — APPROVED** [review]
A little annoying that it's using low level socket but seems like it will do the job.

File: python/s3daemon/s3daemon.py (hunk 4 above)
Comment (ktlim) [reply python/s3daemon/s3daemon.py:76]: Seems extremely unlikely that two pathnames will add up to more than 4KB.  But I can make this unlimited.

File: python/s3daemon/s3daemon.py (hunk 6 above)
Comment (ktlim) [reply python/s3daemon/s3daemon.py:84]: So far this doesn't use any LSST code, and only aiobotocore/botocore outside the Python standard library.  I'd like to keep the dependencies as small as possible.

File: python/s3daemon/s3daemon.py (hunk 7 above)
Comment (ktlim) [reply python/s3daemon/s3daemon.py:83]: Success is only so the client can return a zero exit code.  It doesn't need any more information.

**Pull request merged**